### ISSUE TITLE
Monkey patching with a function of the same signature not supported

### ISSUE THREAD

**Issue by SebbyLaw** [astral-sh/ty#2648]
### Summary

This works in pyright
https://play.ty.dev/fde59fd9-b7d4-4195-a59d-4746fc39acdb
```py
class Foo:
    def method(self) -> None:
        pass

    def add(self, x: int, y: int, /) -> int:
        return x + y


def method_replacement(self) -> None:
    pass


def add_replacement(self: Foo, x: int, y: int, /) -> int:
    return x * y


Foo.method = method_replacement  # error
Foo.add = add_replacement  # error
```

### Version

ty 0.0.14 (16597f5e2 2026-01-26)

**Comment by carljm** [comment]
Thanks for the report! This is a manifestation of #2428; I'll leave it open since it's more discoverable for this particular use case.

**Comment by AlexWaygood** [comment]
> This is a manifestation of [#2428](https://github.com/astral-sh/ty/issues/2428)

Is it? That issue is about the type of methods we infer when they're accessed as attributes on nominal-instance or subclass-of types. But `Foo` here is neither — it's a class-literal type. I assumed we would continue to infer function-literal types when accessing methods as attributes on class-literal types, even after #2428 has been fixed.

**Comment by carljm** [comment]
Good point!

That makes this a little tricky to support. (Pyright and pyrefly allow it; mypy does not.) We could:

1. Also widen the types of methods on a class-literal type to a callable type, allowing this assignment. This might be fine, since to do this widening even for `type[...]` types (#2428) we'll need to track an originating definition for the callable type.
2. Not widen the type on access, but widen it on assignment, in order to allow the monkeypatching. This is a bit weird, but also might be fine?

Implicit in both approaches to supporting this is that go-to-def on the method (and e.g. info sub-diagnostics on bad calls to these methods) would reference the original in-class definition of the method, even though at runtime it might be the monkeypatched method. I think this is also fine, since it's still the in-class definition that defines the type for the method.

**Comment by charliermarsh** [comment]
Should this also apply to class methods and static methods? It's not as straightforward given:

```python
class C:
    @classmethod
    def cm(cls, x: int) -> str: ...

def replacement(cls: type[C], x: int) -> str: ...

C.cm = replacement
C.cm(1)    # Missing an argument
C().cm(1)  # Receives the instance, not C
```

**Comment by carljm** [comment]
I don't think we need to allow this for static/classmethods, at least not in an initial version of it. If we did allow it I think we'd need to require the replacement to also be decorated as a classmethod/staticmethod.

**Comment by charliermarsh** [comment]
Put up that narrow version here: https://github.com/astral-sh/ruff/pull/26158
